proposals/NNNN-lifetime-dependency.md

```diff
@@ -0,0 +1,481 @@
+In the most common cases, these constraints can be inferred automatically.
+
+### Explicit Lifetime Dependency Annotations
+
+To make the semantics clearer, we’ll begin by describing how one can explicitly specify a lifetime constraint in cases where the default inference rules do not apply.
+
+Let’s consider adding support for our hypothetical `BufferReference` type to `Array`.
+Our proposal would allow you to declare an `array.bufferReference()` method as follows:
+
+```swift
+extension Array {
+  borrowing func bufferReference() -> @dependsOn(self) BufferReference<Element> {
+    ... construct a BufferReference ...
+  }
+}
+```
+
+The annotation `@dependsOn(self)` here indicates that the returned value must not outlive the array that produced it.
+Conceptually, it is a continuation of the function's borrowing access:
+the array is being borrowed by the function while the function executes and then continues to be borrowed by the `BufferReference` for as long as the return value exists.
+Specifically, the `@dependsOn(self)` annotation in this example informs the compiler that:
+
+* The array must not be destroyed until after the `BufferReference<Element>` is destroyed.
+  This ensures that use-after-free cannot occur.
+* The array must not be mutated while the  `BufferReference<Element>` value exists.
+  This follows the usual Swift exclusivity rules for a borrowing access.
+
+#### Scoped Lifetime Dependency
+
+Let’s consider another hypothetical type: a `MutatingBufferReference<T>` type that could provide indirect mutating access to a block of memory.
+Here's one way such a value might be produced:
+
+```swift
+func mutatingBufferReference(to: inout Array, count: Int) -> @dependsOn(to) MutatingBufferReference<Element> {
+  ... construct a MutatingBufferReference ...
+}
+```
+
+We’ve written this example as a free function rather than as a method to show how this annotation syntax can be used to express constraints that apply to a particular argument.
+The `@dependsOn(to)` annotation indicates that the returned value depends on the argument named `to`.
+Because `count` is not mentioned in the lifetime dependency, that argument does not participate.
+Similar to the previous example:
+
+* The array will not be destroyed until after the `MutatingBufferReference<Element>` is destroyed.
+* No other read or write access to the array will be allowed for as long as the returned value exists.
+
+In both this and the previous case, the lifetime of the return value is "scoped" to the lifetime of the original value.
+Because lifetime dependencies can only be attached to `~Escapable` values, types that contain pointers will generally need to be `~Escapable` in order to provide safe semantics.
+As a result, **scoped lifetime dependencies** are the common case whenever an `Escapable` value (such as an Array or similar container) is providing a `~Escapable` value (such as the `BufferReference` or `MutatingBufferReference` in these examples).
```

Comment:
**scoped lifetime dependencies** are the common case -> **scoped lifetime dependencies** are the only legal  case

Reply:
Changed.